labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java

```diff
@@ -136,6 +139,59 @@ public void createLabel(
     }
   }
 
+  @Override
+  public void bulkCreateLabels(
+      BulkCreateLabelsRequest bulkCreateLabelsRequest,
+      StreamObserver<BulkCreateLabelsResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      Lock labelsLock = this.stripedLabelsLock.get(requestContext.getTenantId());
+      if (labelsLock.tryLock(WAIT_TIME.getSeconds(), TimeUnit.SECONDS)) {
+        try {
+          List<Label> labels =
+              bulkCreateLabelsRequest.getRequestsList().stream()
+                  .map(this::buildLabelFromRequest)
+                  .filter(Optional::isPresent)
+                  .map(Optional::get)
+                  .collect(Collectors.toList());
+          if (!labels.isEmpty()) {
+            List<Label> createdLabels =
+                labelStore.upsertObjects(requestContext, labels).stream()
+                    .map(org.hypertrace.config.objectstore.ConfigObject::getData)
+                    .collect(Collectors.toList());
+            responseObserver.onNext(
+                BulkCreateLabelsResponse.newBuilder().addAllLabels(createdLabels).build());
+          } else {
+            responseObserver.onNext(BulkCreateLabelsResponse.newBuilder().build());
+          }
+          responseObserver.onCompleted();
+        } finally {
+          labelsLock.unlock();
+        }
+      } else {
+        responseObserver.onError(new StatusRuntimeException(Status.ABORTED));
+      }
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  private Optional<Label> buildLabelFromRequest(BulkCreateLabelsRequest.LabelRequest request) {
+    LabelData labelData = request.getData();
+    if (systemLabelsKeyLabelMap.containsKey(labelData.getKey())
+        || isDuplicateKey(labelData.getKey())) {
```

Comment:
isDuplicateKey does a getAll fetch right? Can we avoid doing that on each request in the bulk create and fetch the exiting ones once?

Reply:
Fixed this part